### PR TITLE
Fix CrossJoinExec evaluating during plan

### LINKED ISSUE #2306: CrossJoin Evaluates In ExecutionPlan::execute
**Describe the bug**

Much like #1939 and #2173, CrossJoin evaluates its children in `ExecutionPlan::execute`

**To Reproduce**

Inspect code

**Expected behavior**

ExecutionPlan::execute should return a stream of results, but should not block on those results being available

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2306

 # Rationale for this change

See ticket

# What changes are included in this PR?

This modifies CrossJoinExec so that it computes the left side data as part of stream execution, and not part of `ExecutionPlan::execute`

# Are there any user-facing changes?

`CrossJoinExec` no longer evaluates during plan

### CODE REVIEW

File: datafusion/core/src/physical_plan/cross_join.rs [hunk 1]
```diff
@@ -352,14 +358,44 @@ impl Stream for CrossJoinStream {
...
+        self.poll_next_impl(cx)
+    }
+}
+
+impl CrossJoinStream {
+    fn poll_next_impl(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<ArrowResult<RecordBatch>>> {
+        let left_result = match &self.left_result {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/cross_join.rs:370]: This is kind of arcane, hopefully the future changes to ExecutionPlan as part of #2199 will make writing these sorts of pipelines easier

File: datafusion/core/src/physical_plan/cross_join.rs [hunk 2]
```diff
@@ -102,6 +106,37 @@ impl CrossJoinExec {
     }
 }
 
+/// Asynchronously collect the result of the left child
+async fn load(
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/cross_join.rs:110]: This is the code that used to be evaluated in `ExecutionPlan::execute`

File: datafusion/core/src/physical_plan/cross_join.rs [hunk 3]
```diff
@@ -352,14 +364,46 @@ impl Stream for CrossJoinStream {
...
+    /// Separate implementation function that unpins the [`CrossJoinStream`] so
+    /// that partial borrows work correctly
+    fn poll_next_impl(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<ArrowResult<RecordBatch>>> {
+        let left_result = match &self.left_result {
+            Some(data) => data,
+            None => {
+                let result = ready!(self.left_fut.poll_unpin(cx));
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/cross_join.rs:381]: Futures-rs does have a MaybeDone construct, but this seemed simpler to understand

File: datafusion/core/src/physical_plan/cross_join.rs [hunk 4]
```diff
@@ -352,14 +364,46 @@ impl Stream for CrossJoinStream {
...
+        let left_data = match left_result.as_ref() {
+            Ok(left_data) => left_data,
+            Err(e) => {
+                return Poll::Ready(Some(Err(ArrowError::ExternalError(
+                    e.to_string().into(),
+                ))))
+            }
+        };
+
+        if left_data.num_rows() == 0 {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/cross_join.rs:395]: I am not sure this is correct in the context of outer joins -- even if the left data has no rows, the stream may still produce output...

However, I see the original code did the same, so 🤷 

```
       if left_data.num_rows() == 0 {
            return Ok(Box::pin(MemoryStream::try_new(
                vec![],
                self.schema.clone(),
                None,
            )?));
        }
```

(it probably only matters for joins that don't have an equality predicate)

**Review by alamb — APPROVED** [review]
LGTM -- thank you @tustvold 

I wish we had some way to test this, but I can't think of anything reasonable at this point

File: datafusion/core/src/physical_plan/cross_join.rs (hunk 3 above)
Comment (tustvold) [reply datafusion/core/src/physical_plan/cross_join.rs:381]: Yup

File: datafusion/core/src/physical_plan/cross_join.rs (hunk 4 above)
Comment (tustvold) [reply datafusion/core/src/physical_plan/cross_join.rs:395]: Yeah, I just blindly reproduced the existing behaviour - I presumed this special case was there for a reason

**Review by andygrove — APPROVED** [review]

**Pull request merged**